javalib/src/main/scala/java/net/URI.scala

```diff
@@ -101,41 +101,37 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
    */
   @inline
   private def internalCompare(that: URI)(cmp: (String, String) => Int): Int = {
-    @inline def cmpOpt(x: js.UndefOr[String], y: js.UndefOr[String]): Int = {
+    @inline
+    def cmpOpt[T](x: js.UndefOr[T], y: js.UndefOr[T])(comparator: (T, T) => Int): Int = {
       if (x == y) 0
       // Undefined components are considered less than defined components
-      else x.fold(-1)(s1 => y.fold(1)(s2 => cmp(s1, s2)))
+      else x.fold(-1)(s1 => y.fold(1)(s2 => comparator(s1, s2)))
     }
 
-    if (this._scheme != that._scheme)
-      this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
-    else if (this._isOpaque != that._isOpaque)
-      // A hierarchical URI is less than an opaque URI
-      if (this._isOpaque) 1 else -1
-    else if (_isOpaque) {
-      val ssp = cmp(this._schemeSpecificPart, that._schemeSpecificPart)
-      if (ssp != 0) ssp
-      else cmpOpt(this._fragment, that._fragment)
-    } else if (this._authority != that._authority) {
-      if (this._host.isDefined && that._host.isDefined) {
-        val ui = cmpOpt(this._userInfo, that._userInfo)
-        if (ui != 0) ui
-        else {
-          val hst = this._host.get.compareToIgnoreCase(that._host.get)
-          if (hst != 0) hst
-          else if (this._port == that._port) 0
-          else if (this._port == -1) -1
-          else if (that._port == -1)  1
-          else this._port - that._port
-        }
-      } else
-        cmpOpt(this._authority, that._authority)
-    } else if (this._path != that._path)
-      cmpOpt(this._path, that._path)
-    else if (this._query != that._query)
-      cmpOpt(this._query, that._query)
-    else
-      cmpOpt(this._fragment, that._fragment)
+    (List(
+      cmpOpt(this._scheme, that._scheme)(_.compareToIgnoreCase(_)),
+      this.isOpaque.compareTo(that.isOpaque), // A hierarchical URI is less than an opaque URI
+    ) ++ (
+      if (this.isOpaque()) {
+        List(
+          cmp(this._schemeSpecificPart, that._schemeSpecificPart),
+        )
+      } else if (this._host.isDefined && that._host.isDefined) {
+        List(
+          cmpOpt(this._userInfo, that._userInfo)(cmp),
+          cmpOpt(this._host, that._host)(_.compareToIgnoreCase(_)),
+          cmpOpt(this._port, that._port)(_ - _),  // change _port to js.UndefOr[Int] and fold into -1 only when getPort is called.
```

Comment:
You actually need to make this change for this to compile :)

Reply:
Oh, I missed.
Fixed in https://github.com/scala-js/scala-js/pull/3690/commits/dfe44551c417702d25aa841379360d442a559ca3